src/extra/bgfx.cpp

```diff
@@ -18,3243 +28,2171 @@
-        const auto key = event.key.keysym.scancode;
-        const auto modState = SDL_GetModState();
-        io.KeysDown[key] = event.type == SDL_KEYDOWN;
-        io.KeyShift = ((modState & KMOD_SHIFT) != 0);
-        io.KeyCtrl = ((modState & KMOD_CTRL) != 0);
-        io.KeyAlt = ((modState & KMOD_ALT) != 0);
-        io.KeySuper = ((modState & KMOD_GUI) != 0);
-      } else if (event.type == SDL_QUIT) {
-        // stop the current chain on close
-        throw ActivationError("Window closed, aborting chain.");
-      } else if (event.type == SDL_WINDOWEVENT &&
-                 SDL_GetWindowID(_window) ==
-                     event.window
-                         .windowID) { // support multiple windows closure
-        if (event.window.event == SDL_WINDOWEVENT_CLOSE) {
-          // stop the current chain on close
-          throw ActivationError("Window closed, aborting chain.");
-        } else if (event.window.event == SDL_WINDOWEVENT_SIZE_CHANGED) {
-          int w, h;
-          SDL_GetWindowSize(_window, &w, &h);
-          CBLOG_DEBUG("GFX.MainWindow size changed width: {} height: {}", w, h);
-        } else if (event.window.event == SDL_WINDOWEVENT_RESIZED) {
-          // get the window size again to ensure it's correct
-          SDL_GetWindowSize(_window, &_rwidth, &_rheight);
-
-          CBLOG_DEBUG("GFX.MainWindow resized width: {} height: {}", _rwidth,
-                      _rheight);
-#ifdef __APPLE__
-          int real_w, real_h;
-          SDL_Metal_GetDrawableSize(_window, &real_w, &real_h);
-          _windowScalingW = float(real_w) / float(_rwidth);
-          _windowScalingH = float(real_h) / float(_rheight);
-          // finally in this case override our real values
-          _rwidth = real_w;
-          _rheight = real_h;
-#endif
-          bgfx::reset(_rwidth, _rheight, BgfxFlags);
-        }
-      }
-    }
+		processWindowEvents(events);
 
-#ifdef __EMSCRIPTEN__
-    double canvasWidth, canvasHeight;
-    EMSCRIPTEN_RESULT res =
-        emscripten_get_element_css_size("#canvas", &canvasWidth, &canvasHeight);
-    if (res != EMSCRIPTEN_RESULT_SUCCESS) {
-      throw ActivationError("Failed to callemscripten_get_element_css_size");
-    }
-    int canvasW = int(canvasWidth);
-    int canvasH = int(canvasHeight);
-    if (canvasW != _rwidth || canvasH != _rheight) {
-      CBLOG_DEBUG("GFX.MainWindow canvas size changed width: {} height: {}",
-                  canvasW, canvasH);
-      SDL_SetWindowSize(_window, canvasW, canvasH);
-      _rwidth = canvasW;
-      _rheight = canvasH;
-      bgfx::reset(_rwidth, _rheight, BgfxFlags);
-    }
-#endif
+		const auto tnow = std::chrono::high_resolution_clock::now();
+		std::chrono::duration<float> ddt = tnow - _deltaTimer.Start;
+		std::chrono::duration<float> adt = tnow - _absTimer.Start;
+		_deltaTimer.Start = tnow;
+
+		gfx::FrameRenderer frameRenderer(_gfxContext, gfx::FrameInputs(ddt.count(), adt.count(), _frameCounter++, events));
```

Comment:
Slow down.. you are doing so many allocations here... It's a huge class.

Reply:
We can optimize later 😉 
Also this is once per frame

Reply:
This is not the way I wrote this codebase until now. Those kind of optimisations should be very preventive.
It's the reason warmup/cleanup/activate exist. And I hope you embrace it.
The whole reason of having such states, chains and blocks is because those kind of optimisations should be intuitive.
Heap allocations should never happen in activate unless it's a one-off that gets cached. 
I think I preached this to @Kryptos-FR as well 😄